### ISSUE TITLE
XGBoost predictions in Python are NaNs when decision class column in pandas dataframe is of type `object`, even though there are only numbers in this column

### ISSUE THREAD

**Issue by mikbuch** [dmlc/xgboost#9715]
I had an issue that I described (along with solution) elsewhere: https://stackoverflow.com/q/77353940/8877692

Below I past the problem description and the solution:

## Problem

have the following problem: for one column of the dataframe the predictions are just fine (i.e., the model output are classes labels, as it should), while for the other column there are only NaNs. There are no NaNs in traning nor prediction "y" vector, so what is causing the problem?

Please note, that this problem is not the same as: [Why H2O XGBoost predicts NaN?](https://stackoverflow.com/questions/65705305/why-h2o-xgboost-predicts-nan) -- as will be explained in the solution below.

## Solution

Well, after some testing and debugging, it turned out that what was causing the problem was the following: for the column (let's call it column Aa) that everything worked correctly -- I just read the data with read_csv, and it was already int-coded. The data in the other column (column B) were in the spreadsheet that I read in a form of stings with literal names of the categories (e.g., cat, dog, etc.). So I had to convert these strings into ints or floats. I did so with the following function:

```
# Need to code the nominal data into number representation (use a dictionary for that).
for class_of_interest in specific_dictionary:
    
    print(class_of_interest)
    
    class_id = specific_dictionary[class_of_interest]
    
    animal_data[animal_data==class_of_interest] = class_id

print(animal_data.unique())
```

(I could not use any pre-defined function, as the categories had to have a specific, dictionary-defined values (e.g., cat=0, dog=1, etc., and the most [dummy] coding/binarizing functions are assigning class labels "at random", probably in the order of occurrence in the column).)

So far, so good. The only thing that I missed is that if I manually code labels like that, the column (Series) type is still an object! I.e., the same as read when the class label was initially a string.

It turns out that in such case XGBoost (and/or scikit-learn) is silently failing because of an object column type (despite there being only numbers in the column), and provides NaNs as prediction output.

If someone know a working coding function that also takes a dictionary as an input, and is additionally changing column data type, I would be grateful for recommending me one.

I leave this question (with an answer) here, as a help for all the people who will be having the same issue in the future. Please note that I am almost 100% sure that this problem is not the same as: "https://stackoverflow.com/questions/65705305/why-h2o-xgboost-predicts-na0" because most probably there the data and/or column data type are causing the issue. If I would have thought differently, I would have posted an answer there. My problem was different, and the solution is data (e.g., H2O) agnostic.

## Proposed solution to the XGBoost codebase

It would be very helpful and informative, if in such situation (i.e., column values are numeric, but column type is object), the framework returned an error (and I mean not even a warning -- a legitimate error message). The rationale is the following: the error would have just pointed to the data type mismatch, and the programmer could correct the error. What is the profit of silently failing and providing the dazzled analytic with NaNs, in such situation?

**Comment by trivialfis** [comment]
Hi, could you please provide an example script that I can run and see the issue so that I have something to test against? XGBoost does some basic checks for dtype: https://github.com/dmlc/xgboost/blob/d4d7097accc4db7d50fdc2b71b643925db6bc424/python-package/xgboost/data.py#L495 . But I assume it's not sufficient for this case, and typing in Python is, hmm, weak.

**Comment by trivialfis** [comment]
Feel free to reopen if there is further info.
